### PR TITLE
Dbutson/python3

### PR DESCRIPTION
**Pull request by dbutson-mdsol**
changes to work with python2 or python3.

### CODE REVIEW

File: bin/nuocd [hunk 1]
```diff
@@ -9,8 +9,16 @@ die () { echo "$*"; exit 1; }
 
 _python=${PYTHON}
 if test -z "$_python"; then
-  _python=$(command -v python2.7) || _python=$(command -v python2) || _python=$(command -v python)  \
-  || die "Python must be installed"
+    _python=$(command -v python3) \
+	|| $(command -v python2.7) \
```
Comment (adriansuarez) [inline bin/nuocd:13]: Shouldn't this be `|| _python="$(command -v python2.7)"`?

File: docker-compose.yml [hunk 2]
```diff
@@ -46,6 +46,7 @@ services:
       INFLUXURL: http://influxdb:8086
       NUOCMD_API_SERVER: nuoadmin1:8888
       NUOCD_HOSTNAME: sm
+      NUOCD_USE_NUOADMIN: "true"
```
Comment (adriansuarez) [inline docker-compose.yml:49]: Not sure how this is still necessary. The non-nuoadmin code should just be removed from nuodb-collector completely and the behavior enabled by `NUOCD_USE_NUOADMIN=true` should be the only option.

File: nuocd/pylib/pynuoadmin/nuocmd.py [hunk 3]
```diff
@@ -0,0 +1,77 @@
+#!/usr/bin/env python
+# (C) Copyright NuoDB, Inc. 2021  All Rights Reserved.
```
Comment (adriansuarez) [inline nuocd/pylib/pynuoadmin/nuocmd.py:2]: This is copying all of pynuoadmin? Why is that necessary, as opposed to just declaring pynuoadmin as a dependency in `requirements.txt`?

File: bin/nuocd (hunk 1 above)
Comment (butson) [reply bin/nuocd:13]: probably, I've been testing by setting the variable PYTHON as I don't have python2 with without python3 installed.  I will confirm.

File: requirements2.txt [hunk 4]
```diff
@@ -0,0 +1,6 @@
+requests >= 2.24.0,<= 2.27.1
+PyYAML >= 5.3.1,<= 5.4.1
+pynuodb >= 2.5.1
+#pynuoadmin >= 2.0.0
```
Comment (adriansuarez) [inline requirements2.txt:4]: Why?

File: requirements3.txt [hunk 5]
```diff
@@ -0,0 +1,6 @@
+requests >= 2.27.1
+PyYAML >= 5.4.1
+pynuodb >= 2.5.1
+#pynuoadmin >= 2.0.0
```
Comment (adriansuarez) [inline requirements3.txt:4]: Why?

File: requirements2.txt (hunk 4 above)
Comment (adriansuarez) [reply requirements2.txt:4]: The latest pynuoadmin is 2.1.0 and supports both Python 2 and Python 3. Is that not usable for some reason?

File: docker-compose.yml (hunk 2 above)
Comment (butson) [reply docker-compose.yml:49]: It's not obvious what NUOCD_USE_NUOADMIN does.  There are two different ways to find nuodb process.   The default was / is to use pgrep nuodb.   The other option is to periodically look for nuodb process on local machine via nuocmd get processes.   That is what this variable does.  I need to confirm.  I ended up merging a change that I did for Nik to eliminate the need to use shared process space between the collector and the engine containers.   I can roll back this change.

File: nuocd/pylib/pynuoadmin/nuocmd.py (hunk 3 above)
Comment (butson) [reply nuocd/pylib/pynuoadmin/nuocmd.py:2]: I wish I could,   the latest pynuoadmin has not been pushed to pypi.  So to get python3 support I must copy until it is pushed.   I would like to remove this but have the dependency on pypi being updated which engineering seems to be incapable of managing.

File: requirements2.txt (hunk 4 above)
Comment (butson) [reply requirements2.txt:4]: see previous comment

File: requirements3.txt (hunk 5 above)
Comment (butson) [reply requirements3.txt:4]: the latest pynuoadmin should be a requirement but it's not available on pypi.

File: nuocd/pylib/pynuoadmin/nuocmd.py (hunk 3 above)
Comment (butson) [reply nuocd/pylib/pynuoadmin/nuocmd.py:2]: Do you have a timeframe when that will be available.   I will close this and wait,  but I want to get this out soon as there are people waiting for it.

File: requirements2.txt [hunk 6]
```diff
@@ -0,0 +1,6 @@
+requests <= 2.24.0
```
Comment (adriansuarez) [inline requirements2.txt:1]: It's also incorrect to say we want requests less than 2.24, because a really outdated unsupported version can satisfy this.

Comment (butson) [reply requirements2.txt:1]: we can just tie it to 2.24.   the >= 2.24 <= 2.27 which seems to be python2 supported versions failed in testing.  <= 2.24.0 will chose 2.24.0 if installing.    I use requirements (on baremetal) to install in our own directory and not in system.  then set PYTHONPATH.

File: requirements3.txt [hunk 7]
```diff
@@ -0,0 +1,7 @@
+testresources==2.0.1
```
Comment (adriansuarez) [inline requirements3.txt:1]: Based on @rshaull's [comment](https://github.com/nuodb/nuodb-collector/issues/23#issuecomment-1353723140), this should not be necessary.

Even if it is due to the cache issue (which he suggested a workaround for), we do not need two requirements files. You can use environment marker to say what the version range should be depending on the Python version. See below:
https://pip.pypa.io/en/stable/reference/requirement-specifiers/

I think the single requirements.txt should have something like:
```
requests >= 2.24.0, < 2.28 ; python_version < '3.0'
requests >= 2.24.0 ; python_version >= '3.0'
```

Comment (butson) [reply requirements3.txt:1]: okay,   I did not know how to do this with one requirements file.   I can rewrite.  makes instructions for deploying easier.

File: requirements2.txt (hunk 4 above)
Comment (adriansuarez) [reply requirements2.txt:4]: I was wrong about 2.1.0 supporting Python 3. A new version of pynuoadmin (2.2.0) will be published that supports Python 3 (>=3.6) and 2.7.

File: nuocd/pylib/pynuoadmin/nuocmd.py (hunk 3 above)
Comment (adriansuarez) [reply nuocd/pylib/pynuoadmin/nuocmd.py:2]: I'm working with release engineering to get it published.

Comment (adriansuarez) [reply nuocd/pylib/pynuoadmin/nuocmd.py:2]: @dbutson-mdsol : Version 2.2.0 of pynuoadmin with Python 3 support has been published: https://pypi.org/project/pynuoadmin/

You can remove this now.

File: nuocd/nuodb_adminquery.py [hunk 8]
```diff
@@ -137,23 +153,39 @@ def get_admin_conn(pid):
...
+
+        if not query_admin:
+            _processes = subprocess.check_output(["pgrep", "^nuodb$"])
+            pids = six.ensure_text(_processes).split()
+        else:
+            # if query admin, only do so if nuodb has not been found.
+            if len(running_local_processes) == 0:
+                pids = [ -1 ]
+            else:
+                pids = [ ]
```
Comment (adriansuarez) [inline nuocd/nuodb_adminquery.py:165]: This query_admin stuff does not seem related to Python 2 vs Python 3. What are these changes and why do we need them for this PR?

File: nuocd/nuodb_adminquery.py [hunk 9]
```diff
@@ -137,23 +153,39 @@ def get_admin_conn(pid):
...
-        _processes = subprocess.check_output(["pgrep", "^nuodb$"])
-        pids = _processes.split()
+
+        if not query_admin:
+            _processes = subprocess.check_output(["pgrep", "^nuodb$"])
+            pids = six.ensure_text(_processes).split()
+        else:
+            # if query admin, only do so if nuodb has not been found.
+            if len(running_local_processes) == 0:
+                pids = [ -1 ]
```
Comment (adriansuarez) [inline nuocd/nuodb_adminquery.py:163]: What is this special -1 pid supposed to signal?

File: nuocd/nuodb_adminquery.py [hunk 10]
```diff
@@ -114,12 +116,26 @@ def get_admin_conn(pid):
                   default=10,
                   type=int,
                   help='interval in seconds between measurements.')
+parser.add_option('-q',
+                  '--query-admin',
+                  dest='query_admin',
+                  default=False,
+                  action='store_true',
+                  help='interval in seconds between measurements.')
```
[the remark this answers is not in the document]
Comment (butson) [reply nuocd/nuodb_adminquery.py:124]: It might be descriptive enough,  that said,  I will be removing it, as this is change relative to fix that is not part of python3 port.  This was accidentally included in this branch.

File: nuocd/nuodb_adminquery.py (hunk 9 above)
Comment (butson) [reply nuocd/nuodb_adminquery.py:163]: I will be removing it, as this is change relative to fix that is not part of python3 port.  This was accidentally included in this branch.  we don't know the pid until the query is done, and we need something in this list to do the query.

Comment (butson) [reply nuocd/nuodb_adminquery.py]: dbutson/poll changes were removed

**Review by adriansuarez — APPROVED** [review]
Looks good.
